middle_end/flambda2/tests/algorithms/patricia_tree_tests.ml

```diff
@@ -0,0 +1,915 @@
+    List.equal Int.equal
+      (l |> Set.of_list |> Set.elements |> List.sort Int.compare)
+      (l |> List.sort_uniq Int.compare)
+
+  let elements_then_of_list s = Set.equal s (s |> Set.elements |> Set.of_list)
+
+  let to_seq s =
+    List.equal Int.equal
+      (s |> Set.to_seq |> List.of_seq |> List.sort Int.compare)
+      (s |> Set.elements |> List.sort Int.compare)
+
+  let union_list l =
+    Set.equal (Set.union_list l) (List.fold_left Set.union Set.empty l)
+
+  let get_singleton s =
+    Option.equal Int.equal (Set.get_singleton s)
+      (match Set.cardinal s with 1 -> Set.choose_opt s | _ -> None)
+end
+
+module Map_specs (V : Value) = struct
+  let ( =? ) = Option.equal V.equal
+
+  let equal_bindings (k1, v1) (k2, v2) = Int.equal k1 k2 && V.equal v1 v2
+
+  let compare_bindings (k1, v1) (k2, v2) =
+    match Int.compare k1 k2 with 0 -> V.compare v1 v2 | c -> c
+
+  let find_opt_vs_find k m =
+    Map.find_opt k m =? option_of_not_found (Map.find k) m
+
+  let find_opt_empty k = Map.find_opt k Map.empty =? None
+
+  let mem k m = Map.find_opt k m =? None <=> not (Map.mem k m)
+
+  let is_empty_vs_equal m = Map.is_empty m <=> Map.equal V.equal m Map.empty
+
+  let add_valid k v m = Map.valid (Map.add k v m)
+
+  let add_same k v m = Map.find_opt k (Map.add k v m) =? Some v
+
+  let add_other k1 k2 v m =
+    k1 = k2 || Map.find_opt k1 (Map.add k2 v m) =? Map.find_opt k1 m
+
+  let update_valid k f m = Map.valid (Map.update k f m)
+
+  let update_same k f m =
+    Map.find_opt k (Map.update k f m) =? f (Map.find_opt k m)
+
+  let update_other k1 k2 f m =
+    k1 = k2 || Map.find_opt k1 (Map.update k2 f m) =? Map.find_opt k1 m
+
+  let singleton_valid k v = Map.valid (Map.singleton k v)
+
+  let singleton k1 k2 v =
+    Map.find_opt k1 (Map.singleton k2 v) =? if k1 = k2 then Some v else None
+
+  let remove_valid k m = Map.valid (Map.remove k m)
+
+  let remove_same k m = Map.find_opt k (Map.remove k m) =? None
+
+  let remove_other k1 k2 m =
+    k1 = k2 || Map.find_opt k1 (Map.remove k2 m) =? Map.find_opt k1 m
+
+  let merge_valid f m1 m2 = Map.valid (Map.merge f m1 m2)
+
+  let merge f m1 m2 =
+    let mm = Map.merge f m1 m2 in
+    let correct_for_key k =
+      Map.find_opt k mm =? f k (Map.find_opt k m1) (Map.find_opt k m2)
+    in
+    Map.for_all (fun k _ -> correct_for_key k) m1
+    && Map.for_all (fun k _ -> correct_for_key k) m2
+
+  let union_valid f m1 m2 = Map.valid (Map.union f m1 m2)
+
+  let union f m1 m2 =
+    let mu = Map.union f m1 m2 in
+    m1
+    |> Map.for_all (fun k v1 ->
+           match Map.find_opt k m2 with
+           | None -> Map.find_opt k mu =? Some v1
+           | Some v2 -> Map.find_opt k mu =? f k v1 v2)
+    && m2
+       |> Map.for_all (fun k v2 ->
+              match Map.find_opt k m1 with
+              | None -> Map.find_opt k mu =? Some v2
+              | Some _ -> (* already checked *) true)
+    && mu |> Map.for_all (fun k _ -> Map.mem k m1 || Map.mem k m2)
```

Comment:
Lots of conjunctions – I'm not sure you want to split them, but it's the sort of place where that could be good.

Reply:
Yeah, I split off the third one.